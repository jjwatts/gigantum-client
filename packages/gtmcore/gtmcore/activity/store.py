# Copyright (c) 2017 FlashX, LLC
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.
import re
import uuid
import datetime
from typing import (Any, Dict, List, Tuple, Optional)

from gtmcore.activity.detaildb import ActivityDetailDB
from gtmcore.activity.records import ActivityDetailRecord, ActivityRecord
from gtmcore.logging import LMLogger

logger = LMLogger.get_logger()


class ActivityStore(object):
    """The ActivityStore class provides a centralized interface to activity data stored in both the git log and db.

    High-level information is stored directly in the git commit messages stored in the git log of a repo. For more
    detailed information and arbitrary information a custom checkout-aware file based key-value store is used.

    The ActivityStore class stores ActivityRecords in the git log and ActivityDetailRecords in the database after
    proper serialization. The linked commit indiciate which git record the ActivityRecord is annotating
    """

    def __init__(self, repository) -> None:
        """ Load the database for the specified repository

        Args:
            repository(gtmcore.inventory.repository.Repository): A Repository instance
        """
        # Configuration parameters
        self.max_num_tags: int = 100
        self.max_tag_length: int = 256

        self.repository = repository

        self.detaildb = ActivityDetailDB(repository.root_dir, repository.checkout_id,
                                         logfile_limit=repository.client_config.config['detaildb']['logfile_limit'])

        # Note record commit messages follow a special structure
        self.note_regex = re.compile(r"(?s)_GTM_ACTIVITY_START_.*?_GTM_ACTIVITY_END_")

        # Params used during detail object serialization
        if self.repository.client_config.config['detaildb']['options']['compress']:
            self.compress_details: bool = self.repository.client_config.config['detaildb']['options']['compress']
            self.compress_min_bytes = self.repository.client_config.config['detaildb']['options']['compress_min_bytes']
        else:
            self.compress_details: bool = False
            self.compress_min_bytes: int = 0

    def _validate_tags(self, tags: List[str]) -> List[str]:
        """Method to clean and validate tags

        Args:
            tags(list): A list of strings

        Returns:
            (list): a list of strings
        """
        # allow for no tags
        if not tags:
            return []

        # Remove duplicate tags
        tags = list(set(tags))

        # Check total number of tags
        if len(tags) > self.max_num_tags:
            raise ValueError(
                f"{len(tags)} tags provided, but a single Activity Record can only have {self.max_num_tags} tags.")

        # Check tag length
        for tag in tags:
            if len(tag) > self.max_tag_length:
                msg = "tag `{}` has {} characters,".format(tag, len(tag))
                raise ValueError("{} but a  tag is limited to {} characters.".format(msg, self.max_tag_length))

        # Remove \`; as a very basic level of sanitization
        return [tag.strip().translate({ord(c): None for c in '\`;'}) for tag in tags]

    def _get_log_records(self, after: Optional[str]=None, before: Optional[str]=None,
                         first: Optional[int]=None,
                         last: Optional[int]=None) -> List[Tuple[str, str, datetime.datetime, str, str]]:
        """Method to get ACTIVITY records from the git log with pagination support

        Returns:
            list: List of tuples of the format (log string, commit hash, commit datetime, username, email)
        """
        log_entries: List[Tuple[str, str, datetime.datetime, str, str]] = list()
        kwargs = dict()

        # TODO: Add support for reverse paging
        if before:
            raise ValueError("Paging using the 'before' argument not yet supported.")
        if last:
            raise ValueError("Paging using the 'last' argument not yet supported.")

        if first is not None:
            if first < 1:
                raise ValueError("`first` must be greater than or equal to 1, or None")

            # We typically have 2 commits per activity, 1 for the actual user changes and 1 for our changes.
            # To page properly, load up to 2x the number requested plus 5 to be safe in most cases
            kwargs['max_count'] = (first * 2) + 5

        path_info: Optional[str] = None
        if after:
            path_info = after

        while True:
            for entry in self.repository.git.log(path_info=path_info, **kwargs):
                m = self.note_regex.match(entry['message'])
                if m:
                    log_entries.append((m.group(0), entry['commit'], entry['committed_on'],
                                        entry['author']['name'],
                                        entry['author']['email']))

            if first is not None:
                if first == -1:
                    # If you get here, you already tried to load more records. Give up
                    break

                elif after is not None and len(log_entries) < first + 1:
                    # Didn't get enough records back. Possibly an edge case to handle or possibly just the end
                    # Try getting lots of records (paging here is mainly for performance reasons only)
                    kwargs['max_count'] = 100

                    # Set first to -1 to ensure you don't keep looping at the end of a git log
                    first = -1
                    log_entries = list()

                elif len(log_entries) < first:
                    # Didn't get enough records back. Possibly an edge case to handle or possibly just the end
                    # Try getting lots of records (paging here is mainly for performance reasons only)
                    kwargs['max_count'] = 100

                    # Set first to -1 to ensure you don't keep looping at the end of a git log
                    first = -1
                    log_entries = list()

                else:
                    break
            else:
                break

        return log_entries

    def create_activity_record(self, record: ActivityRecord) -> ActivityRecord:
        """Method to write an activity record and its details to the git log and detaildb

        Args:
            record(ActivityRecord): A populated activity record

        Returns:
            ActivityRecord
        """
        # If there isn't a linked commit, generate a UUID to uniquely ID the data in levelDB that will never
        # collide with the actual git hash space by making it 32 char vs. 40 for git
        if not record.linked_commit:
            record.linked_commit = uuid.uuid4().hex

        # Write all ActivityDetailObjects to the datastore
        with record.inspect_detail_objects() as details:
            for idx, detail in enumerate(details):
                # TODO DC: Do we want to try and put some error handling here to ignore individual detail records?
                # E.g., set directive to 'ignore'?
                # I think that an exception may result in a malformed detaildb / activity record
                updated_detail = self.put_detail_record(detail)
                record.update_detail_object(updated_detail, idx)

        # Add everything in the repo activity/log directory
        self.repository.git.add_all(self.detaildb.root_path)

        # Commit changes and update record
        commit = self.repository.git.commit(record.log_str)
        record.commit = commit.hexsha

        # Update record with username and email
        record.username = self.repository.git.author.name
        record.email = self.repository.git.author.email

        logger.debug(f"Successfully created ActivityRecord {commit.hexsha}")
        return record

    def get_activity_record(self, commit: str) -> ActivityRecord:
        """Method to get a single ActivityRecord

        Args:
            commit(str): The commit hash of the activity record

        Returns:
            ActivityRecord
        """
        entry = self.repository.git.log_entry(commit)
        m = self.note_regex.match(entry["message"])
        if m:
            ar = ActivityRecord.from_log_str(m.group(0), commit, entry['committed_on'],
                                             username=entry['author']['name'], email=entry['author']['email'])
            return ar
        else:
            raise ValueError("Activity data not found in commit {}".format(commit))

    def get_activity_records(self, after: Optional[str]=None,
                             first: Optional[int]=None) -> List[Optional[ActivityRecord]]:
        """Method to get a list of activity records, with forward paging supported

        Args:
            after(str): Commit hash to page after
            first(int): Number of records to get

        Returns:
            List[ActivityRecord]
        """
        # Get data from the git log
        log_data = self._get_log_records(after=after, first=first)
        if log_data:
            if after:
                # If the "after" record is included. Remove it due to standards on how relay paging works
                log_data = log_data[1:]

            # If first value provided, check for the right amount of data
            if first:
                if len(log_data) > first:
                    # Need to prune due to padding sent into self._get_log_records()
                    log_data = log_data[:first]

        if log_data:
            return [ActivityRecord.from_log_str(x[0], x[1], x[2], username=x[3], email=x[4]) for x in log_data]
        else:
            return []

    def _encode_write_options(self, compress: bool = False) -> bytes:
        """Method to encode any options for writing details to a byte

        bit option
        0   compress/decompress data on storage
        1   reserved
        2   reserved
        3   reserved
        4   reserved
        5   reserved
        6   reserved
        7   reserved

        Returns:
            bytes
        """
        return compress.to_bytes(1, byteorder='little')

    @staticmethod
    def _decode_write_options(option_byte: bytes) -> dict:
        """Method to decode the write options header byte

        Args:
            option_byte(bytes): Byte containing option flags

        Returns:
            dict
        """
        return {"compress": bool(option_byte[0])}

    def put_detail_record(self, detail_obj: ActivityDetailRecord) -> ActivityDetailRecord:
        """Method to write a detail record to the activity detail db

        Args:
            detail_obj(ActivityDetailRecord): The detail record to write

        Returns:
            ActivityDetailRecord: the detail record updated with the key
        """
        # Set compression option based on config and objects size
        compress = False
        if self.compress_details:
            if detail_obj.data_size >= self.compress_min_bytes:
                compress = True

        bytes_record = detail_obj.to_bytes(compress)

        # Write record and store key
        detail_obj.key = self.detaildb.put(self._encode_write_options(compress=compress) +
                                           bytes_record)

        logger.debug(f"Successfully wrote ActivityDetailRecord {detail_obj.key}")
        return detail_obj

    def get_detail_record(self, detail_key: str) -> ActivityDetailRecord:
        """Method to fetch a detail entry from the activity detail db

            Args:
                detail_key : the key returned from the activity detail DB when storing.

            Returns:
                 ActivityDetailRecord
        """
        # Get value from key-value store
        detail_bytes = self.detaildb.get(detail_key)

        # Remove header
        options = self._decode_write_options(detail_bytes[:1])

        # Create object
        record = ActivityDetailRecord.from_bytes(detail_bytes[1:], decompress=options['compress'])
        record.key = detail_key
        return record
