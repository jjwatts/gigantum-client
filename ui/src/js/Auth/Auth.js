import history from 'JS/history';
import auth0 from 'auth0-js';
import RemoveUserIdentityMutation from 'Mutations/RemoveUserIdentityMutation';
// queries
import SessionCheck from 'JS/Auth/sessionCheck';
// store
import { setLogout, setLoginError } from 'JS/redux/actions/login';
// variables
import { AUTH_CONFIG } from './auth0-variables';

export default class Auth {
  auth0 = new auth0.WebAuth({
    domain: AUTH_CONFIG.domain,
    clientID: AUTH_CONFIG.clientId,
    redirectUri: AUTH_CONFIG.callbackUrl,
    audience: AUTH_CONFIG.audience,
    responseType: 'token id_token',
    scope: 'openid profile email user_metadata',
  });

  constructor() {
    this.login = this.login.bind(this);
    this.logout = this.logout.bind(this);
    this.handleAuthentication = this.handleAuthentication.bind(this);
    this.isAuthenticated = this.isAuthenticated.bind(this);
    this.renewToken = this.renewToken.bind(this);
  }

  /**
   * Renews auth token if possible, otherwise prompt login
  */
  renewToken(showModal, showModalCallback, successCallback, forceHistory, failureCallback) {
    this.auth0.checkSession({}, (err, result) => {
      if (err) {
        if (showModal) {
          showModalCallback();
        } else {
          failureCallback();
        }
      } else {
        this.setSession(result, true, forceHistory);
        if (successCallback) {
          successCallback();
        }
      }
    });
  }

  login() {
    setLogout(false);
    this.auth0.authorize();
  }

  handleAuthentication() {
    this.auth0.parseHash((err, authResult) => {
      if (authResult && authResult.accessToken && authResult.idToken) {
        this.setSession(authResult);
        window.sessionStorage.removeItem('LOGIN_ERROR_DESCRIPTION');
        window.sessionStorage.removeItem('LOGIN_ERROR_TYPE');
      } else if (err) {
        history.replace('/login');
        setLoginError(err);
        window.sessionStorage.setItem('LOGIN_ERROR_TYPE', err.error);
        window.sessionStorage.setItem('LOGIN_ERROR_DESCRIPTION', err.errorDescription);
        //  alert(`Error: ${err.error}. Check the console for further details.`); TODO make this a modal or redirect to login failure page
      }
    });
  }

  setSession(authResult, silent, forceHistory) {
    // Set the time that the access token will expire at
    const expiresAt = JSON.stringify((authResult.expiresIn * 1000) + new Date().getTime());
    localStorage.setItem('access_token', authResult.accessToken);
    localStorage.setItem('id_token', authResult.idToken);
    localStorage.setItem('expires_at', expiresAt);
    localStorage.setItem('family_name', authResult.idTokenPayload.family_name);
    localStorage.setItem('given_name', authResult.idTokenPayload.given_name);
    localStorage.setItem('email', authResult.idTokenPayload.email);
    localStorage.setItem('username', authResult.idTokenPayload.nickname);
    // redirect to labbooks when user logs in
    let route = window.sessionStorage.getItem('CALLBACK_ROUTE')
      ? window.sessionStorage.getItem('CALLBACK_ROUTE')
      : '/projects';

    route = route === ''
      ? '/projects'
      : route;

    if (!silent || forceHistory) {
      history.replace(route);
    }
  }

  logout() {
    setLogout(true);
    RemoveUserIdentityMutation(() => {
      // redirect to root when user logs out

      localStorage.removeItem('access_token');
      localStorage.removeItem('id_token');
      localStorage.removeItem('expires_at');
      localStorage.removeItem('family_name');
      localStorage.removeItem('given_name');
      localStorage.removeItem('email');
      localStorage.removeItem('username');
      window.sessionStorage.removeItem('CALLBACK_ROUTE');

      history.replace('/');
    });
  }

  isAuthenticated() {
    // Check whether the current time is past the
    // access token's expiry time

    return SessionCheck.getUserIdentity().then(response => response && response.data && response.data.userIdentity && response.data.userIdentity.isSessionValid);
  }
}
