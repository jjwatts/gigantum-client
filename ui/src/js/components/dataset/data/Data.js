// vendor
import { createFragmentContainer, graphql } from 'react-relay';
// component
import SectionWrapper from 'Components/shared/filesShared/sectionWrapper/SectionWrapper';

export default createFragmentContainer(
  SectionWrapper,
  graphql`
      fragment Data_dataset on Dataset{
        id @skip (if: $dataSkip)
        ...DataBrowser_data @skip (if: $dataSkip)
      }
    `,
);
