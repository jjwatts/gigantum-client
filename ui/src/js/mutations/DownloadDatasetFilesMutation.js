import {
  commitMutation,
  graphql,
} from 'react-relay';
import environment from 'JS/createRelayEnvironment';
// utils
import FooterUtils from 'Components/common/footer/FooterUtils';
import footerCallback from 'Components/common/footer/utils/DownloadDatasetFiles';
import { setErrorMessage } from 'JS/redux/actions/footer';

const mutation = graphql`
mutation DownloadDatasetFilesMutation($input: DownloadDatasetFilesInput!){
    downloadDatasetFiles(input: $input){
        backgroundJobKey
        clientMutationId
    }
}
`;

let tempID = 0;


export default function DownloadDatasetFilesMutation(
  datasetOwner,
  datasetName,
  labbookName,
  labbookOwner,
  successCall,
  failureCall,
  keys,
  allKeys,
  callback,
) {
  const variables = {
    input: {
      datasetOwner,
      datasetName,
      clientMutationId: tempID++,
    },
  };
  if (allKeys) {
    variables.input.allKeys = allKeys;
  } else {
    variables.input.keys = keys;
  }
  if (labbookName) {
    variables.input.labbookName = labbookName;
  }
  if (labbookOwner) {
    variables.input.labbookOwner = labbookOwner;
  }

  commitMutation(environment, {
    mutation,
    variables,
    onCompleted: (response, error) => {
      if (error) {
        setErrorMessage('There was a problem downloading Dataset files', error);
        console.log(error);
      }
      const footerData = {
        result: response,
        type: 'downloadDatasetFiles',
        key: 'backgroundJobKey',
        footerCallback,
        successCall,
        failureCall,
      };

      FooterUtils.getJobStatus(footerData);

      callback(error);
    },
    onError: err => console.error(err),
  });
}
