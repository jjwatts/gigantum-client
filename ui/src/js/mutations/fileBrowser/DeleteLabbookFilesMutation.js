import {
  commitMutation,
  graphql,
} from 'react-relay';
import RelayRuntime from 'relay-runtime';
import environment from 'JS/createRelayEnvironment';


const mutation = graphql`
  mutation DeleteLabbookFilesMutation($input: DeleteLabbookFilesInput!){
    deleteLabbookFiles(input: $input){
      success
      clientMutationId
    }
  }
`;

let tempID = 0;

export default function DeleteLabbookFilesMutation(
  connectionKey,
  owner,
  labbookName,
  labbookId,
  filePaths,
  section,
  edgesToDelete,
  callback,
) {
  const variables = {
    input: {
      owner,
      labbookName,
      filePaths,
      section,
      clientMutationId: `${tempID++}`,
    },
  };

  function sharedUpdater(store, labbookID, deletedID, connectionKey) {
    const userProxy = store.get(labbookID);

    const conn = RelayRuntime.ConnectionHandler.getConnection(
      userProxy,
      connectionKey,
    );

    if (conn) {
      RelayRuntime.ConnectionHandler.deleteNode(
        conn,
        deletedID,
      );
    }

    if (store.get(deletedID)) {
      store.delete(deletedID);
    }
  }

  commitMutation(
    environment,
    {
      mutation,
      variables,
      configs: [{
        type: 'NODE_DELETE',
        connectionKeys: [{
          key: connectionKey,
        }],
        parentId: labbookId,
        pathToConnection: ['labbook', 'allFiles'],
      }],
      onCompleted: (response, error) => {
        if (error) {
          console.log(error);
        }
        callback(response, error);
      },
      onError: err => console.error(err),

      updater: (store) => {
        edgesToDelete.forEach((edge) => {
          if (edge) {
            sharedUpdater(store, labbookId, edge.node.id, connectionKey);

            store.delete(edge.node.id);
          }
        });
      },
      optimisticUpdater: (store) => {
        edgesToDelete.forEach((edge) => {
          if (edge) {
            sharedUpdater(store, labbookId, edge.node.id, connectionKey);
            store.delete(edge.node.id);
          }
        });
      },
    },
  );
}
