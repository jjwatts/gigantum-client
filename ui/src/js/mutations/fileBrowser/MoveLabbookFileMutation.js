import {
  commitMutation,
  graphql,
} from 'react-relay';
import RelayRuntime from 'relay-runtime';
import environment from 'JS/createRelayEnvironment';

import { setErrorMessage } from 'JS/redux/actions/footer';

const mutation = graphql`
  mutation MoveLabbookFileMutation($input: MoveLabbookFileInput!){
    moveLabbookFile(input: $input){
      updatedEdges{
        node{
            id
            isDir
            modifiedAt
            key
            size
          }
          cursor
      }
      clientMutationId
    }
  }
`;

let tempID = 0;

function sharedDeleteUpdater(store, labbookID, removeIds, connectionKey) {
  const labbookProxy = store.get(labbookID);
  if (labbookProxy) {
    removeIds.forEach((deletedID) => {
      const conn = RelayRuntime.ConnectionHandler.getConnection(
        labbookProxy,
        connectionKey,
      );

      if (conn) {
        RelayRuntime.ConnectionHandler.deleteNode(
          conn,
          deletedID,
        );
        store.delete(deletedID);
      }
    });
  }
}


export default function MoveLabbookFileMutation(
  connectionKey,
  owner,
  labbookName,
  labbookId,
  edge,
  srcPath,
  dstPath,
  section,
  removeIds,
  callback,
) {
  const variables = {
    input: {
      owner,
      labbookName,
      srcPath,
      dstPath,
      section,
      clientMutationId: `${tempID++}`,
    },
  };

  const configs = [{
    type: 'RANGE_ADD',
    parentID: labbookId,
    connectionInfo: [{
      key: connectionKey,
      rangeBehavior: 'append',
    }],
    edgeName: 'newLabbookFileEdge',
  }];

  if (removeIds && removeIds.length) {
    removeIds.forEach((id) => {
      configs.unshift({
        type: 'NODE_DELETE',
        deletedIDFieldName: id,
      });
    });
  }

  commitMutation(
    environment,
    {
      mutation,
      variables,
      onCompleted: (response, error) => {
        if (error) {
          console.log(error);
          setErrorMessage(`ERROR: Could not Move labbook file ${srcPath}`, error);
        }
        callback(response, error);
      },
      onError: err => console.error(err),
      optimisticUpdater: (store) => {
        const labbookProxy = store.get(labbookId);

        if (labbookProxy && (edge.node !== null)) {
          const conn = RelayRuntime.ConnectionHandler.getConnection(
            labbookProxy,
            connectionKey,
          );

          const node = store.get(edge.node.id);
          node.setValue(edge.node.id, 'id');
          node.setValue(edge.node.isDir, 'isDir');
          node.setValue(dstPath, 'key');
          node.setValue(edge.node.modifiedAt, 'modifiedAt');
          node.setValue(edge.node.size, 'size');

          if (!store.get(edge.node.id)) {
            const newEdge = RelayRuntime.ConnectionHandler.createEdge(
              store,
              conn,
              node,
              'newLabbookFileEdge',
            );
            RelayRuntime.ConnectionHandler.insertEdgeAfter(
              conn,
              newEdge,
              edge.cursor,
            );
          }
        }
      },
      updater: (store, response) => {
        sharedDeleteUpdater(store, labbookId, removeIds, connectionKey);
        if (response && response.moveLabbookFile && response.moveLabbookFile.updatedEdges) {
          response.moveLabbookFile.updatedEdges.forEach((edge) => {
            const labbookProxy = store.get(labbookId);

            if (labbookProxy && (edge.node !== null)) {
              const conn = RelayRuntime.ConnectionHandler.getConnection(
                labbookProxy,
                connectionKey,
              );
              const node = store.get(edge.node.id) ? store.get(edge.node.id) : store.create(edge.node.id, 'LabbookFile');

              node.setValue(edge.node.id, 'id');
              node.setValue(edge.node.isDir, 'isDir');
              node.setValue(edge.node.key, 'key');
              node.setValue(edge.node.modifiedAt, 'modifiedAt');
              node.setValue(edge.node.size, 'size');
              const newEdge = RelayRuntime.ConnectionHandler.createEdge(
                store,
                conn,
                node,
                'newLabbookFileEdge',
              );
              RelayRuntime.ConnectionHandler.insertEdgeAfter(
                conn,
                newEdge,
                edge.cursor,
              );
            }
          });
        }
      },

    },
  );
}
