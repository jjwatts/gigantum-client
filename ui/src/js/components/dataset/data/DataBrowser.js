// vendor
import { createPaginationContainer, graphql } from 'react-relay';
// component
import SectionBrowser from 'Components/shared/filesShared/sectionBrowser/SectionBrowser';
// store
import store from 'JS/redux/store';

export default createPaginationContainer(
  SectionBrowser,
  {

    data: graphql`
        fragment DataBrowser_data on Dataset{
          allFiles(after: $cursor, first: $first)@connection(key: "DataBrowser_allFiles", filters: []){
            edges{
              node{
                id
                isDir
                isLocal
                modifiedAt
                key
                size
              }
              cursor
            }
            pageInfo{
              hasNextPage
              hasPreviousPage
              startCursor
              endCursor
            }
          }
        }`,
  },
  {
    direction: 'forward',
    getConnectionFromProps(props) {
      return props.data && props.data.allFiles;
    },
    getFragmentVariables(prevVars, totalCount) {
      return {
        ...prevVars,
        first: totalCount,
      };
    },
    getVariables(props, { count, cursor }, fragmentVariables) {
      const { owner, labbookName } = store.getState().routes;

      return {
        first: count,
        cursor,
        owner,
        name: labbookName,
      };
    },
    query: graphql`
        query DataBrowserPaginationQuery(
          $first: Int
          $cursor: String
          $owner: String!
          $name: String!
        ) {
          dataset(name: $name, owner: $owner){
             id
             description
             ...DataBrowser_data
          }
        }
      `,
  },

);
