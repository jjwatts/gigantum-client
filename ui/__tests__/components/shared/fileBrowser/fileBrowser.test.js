// vendor
import { Provider } from 'react-redux';
import React from 'react';
import renderer from 'react-test-renderer';
import FileBrowserDropZone from 'Components/shared/fileBrowser/FileBrowser';
import { mount, shallow } from 'enzyme';
import { BrowserRouter as Router } from 'react-router-dom';
// import { Provider } from 'react-redux';
import store from 'JS/redux/store';
// fixtures
import Auth from 'JS/Auth/Auth';
import history from 'JS/history';
// data
import data from 'Tests/components/labbook/code/__relaydata__/MostRecentCode.json';
import codeData from 'Tests/components/labbook/code/__relaydata__/CodeBrowser.json';
import codeDataFavorites from 'Tests/components/labbook/code/__relaydata__/CodeFavorites.json';
// mutations
import DeleteLabbookFilesMutation from 'Mutations/fileBrowser/DeleteLabbookFilesMutation';
import MakeLabbookDirectoryMutation from 'Mutations/fileBrowser/MakeLabbookDirectoryMutation';
import MoveLabbookFileMutation from 'Mutations/fileBrowser/MoveLabbookFileMutation';
import DownloadDatasetFilesMutation from 'Mutations/DownloadDatasetFilesMutation';
import CompleteBatchUploadTransactionMutation from 'Mutations/fileBrowser/CompleteBatchUploadTransactionMutation';

const auth = new Auth();


const variables = { first: 20, labbook: 'ui-test-project' };
export default variables;

const { labbook } = codeData.data;

const setRootFolder = () => {};

const fixtures = {
  connectDropTarget: jsx => jsx,
  clearSelectedFiles: jest.fn(),
  loadStatus: jest.fn(),
  setRootFolder: jest.fn(),
  selectedFiles: [],
  labbookId: labbook.id,
  sectionId: labbook.code.id,
  section: 'code',
  isLocked: false,
  ...labbook.code,
  labbook,
  files: labbook.code.allFiles,
  parentId: labbook.id,
  connection: 'CodeBrowser_allFiles',
  favoriteConnection: 'CodeFavorites_favorites',
  favorites: codeDataFavorites.data.labbook.code.favorites,
};

jest.mock('Mutations/fileBrowser/DeleteLabbookFilesMutation', () => jest.fn());
jest.mock('Mutations/fileBrowser/MakeLabbookDirectoryMutation', () => jest.fn());
jest.mock('Mutations/fileBrowser/MoveLabbookFileMutation', () => jest.fn());
jest.mock('Mutations/DownloadDatasetFilesMutation', () => jest.fn());
jest.mock('Mutations/fileBrowser/CompleteBatchUploadTransactionMutation', () => jest.fn());


auth.isAuthenticated = function () { return true; };
const FileBrowser = FileBrowserDropZone.DecoratedComponent;

describe('FileBrowser component', () => {
  it('Test FileBrowser Rendering', () => {
        const component = renderer.create(<Provider store={store}><FileBrowserDropZone.DecoratedComponent
              {...fixtures}
            /></Provider>);

        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
   });

   const component = mount(<FileBrowserDropZone.DecoratedComponent {...fixtures } />);

   let deleteCount = 0;
   const MockFn = jest.fn();
   const mockDeleteMutation = new MockFn();
   component._deleteMutation = mockDeleteMutation;

   it('Sorts by date', () => {
     component.find('.FileBrowser__header--date').simulate('click');
     expect(component.state().sort).toEqual('modified');
   });

   it('Multiselect selects all', () => {
     component.find('.CheckboxMultiselect').simulate('click');
     expect(component.state('multiSelect')).toEqual('all');
   });

   it('Cancel delete popup', () => {
     component.find('.File__btn--delete').simulate('click');
     expect(component.state('popupVisible')).toEqual(false);
   });

   it('Updates search input', () => {
     const evt = {
       target: {
         value: 'png',
       },
     };
     component.find('.FileBrowser__input').simulate('change', evt);
     expect(component.state('search')).toEqual('png');
   });


   it('Sorts by az', () => {
     component.find('.FileBrowser__header--name').simulate('click');
     expect(component.state('sort')).toEqual('az');
   });

   it('Sorts by size', () => {
     component.find('.FileBrowser__header--size').simulate('click');
     expect(component.state('sort')).toEqual('size');
   });

   it('Sorts by modified', () => {
     component.find('.FileBrowser__header--date').simulate('click');
     expect(component.state('sort')).toEqual('modified');
   });

});
