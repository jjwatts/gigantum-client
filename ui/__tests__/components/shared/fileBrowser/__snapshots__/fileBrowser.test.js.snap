// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileBrowser component Test FileBrowser Rendering 1`] = `
<div
  className="FileBrowser FileBrowser--dropzone"
  style={
    Object {
      "zIndex": 0,
    }
  }
>
  <div
    className="FileBrowser__menu"
  >
    <h5>
      Files
    </h5>
    <div
      className="FileBrowser__menu-buttons"
    >
      <div>
        <label
          className="FileBrowser__upload-label"
          htmlFor="browser_upload"
        >
          <div
            className="Btn--fileBrowser Btn--round Btn--bordered Btn__upArrow inline-block Btn"
          />
          <span>
            Upload Files
          </span>
          <input
            className="hidden"
            id="browser_upload"
            multiple={true}
            onChange={[Function]}
            type="file"
          />
        </label>
        <button
          className="Btn Btn__menuButton Btn--noShadow FileBrowser__newFolder"
          data-click-id="addFolder"
          onClick={[Function]}
          type="button"
        >
          <div
            className="Btn--fileBrowser Btn--round Btn--bordered Btn__addFolder Btn"
            data-click-id="addFolder"
          />
          New Folder
        </button>
      </div>
    </div>
  </div>
  <div
    className="FileBrowser__tools flex justify--space-between"
  >
    <div
      className="FileBrowser__search flex-1"
    >
      <input
        className="FileBrowser__input search"
        onChange={[Function]}
        onKeyUp={[Function]}
        placeholder="Search Files Here"
        type="text"
      />
    </div>
  </div>
  <div
    className="FileBrowser__header"
  >
    <div
      className="FileBrowser__multiselect flex justify--start"
    >
      <button
        className="CheckboxMultiselect CheckboxMultiselect__uncheck"
        onClick={[Function]}
        type="button"
      />
      <button
        className="Btn Btn--fileBrowser Btn--round Btn__delete Btn--action hidden"
        onClick={[Function]}
        type="button"
      />
      <div
        className="FileBrowser__popup hidden Tooltip__message"
      >
        <div
          className="Tooltip__pointer"
        />
        <p>
          Are you sure?
        </p>
        <div
          className="flex justify--space-around"
        >
          <button
            className="File__btn--round File__btn--cancel File__btn--delete"
            onClick={[Function]}
            type="button"
          />
          <button
            className="File__btn--round File__btn--add File__btn--delete-files"
            onClick={[Function]}
            type="button"
          />
        </div>
      </div>
    </div>
    <button
      className="FileBrowser__name-text FileBrowser__header--name flex justify--start Btn--noStyle FileBroser__sort--asc"
      onClick={[Function]}
      type="button"
    >
      File
    </button>
    <button
      className="FileBrowser__header--size Btn--noStyle"
      onClick={[Function]}
      type="button"
    >
      Size
    </button>
    <button
      className="FileBrowser__header--date Btn--noStyle"
      onClick={[Function]}
      type="button"
    >
      Modified
    </button>
    <div
      className="FileBrowser__header--menu flex flex--row justify--right"
    />
  </div>
  <div
    className="FileBrowser__body"
  >
    <div
      className="AddSubfolder hidden"
      onClick={[Function]}
    >
      <div
        className="AddSubfolder__text"
      >
        Add Folder
      </div>
      <div
        className="AddSubfolder__edit Input--clear hidden"
      >
        <div
          className="AddSubfolder__container"
        >
          <input
            className="AddSubfolder__input"
            onKeyUp={[Function]}
            placeholder="Enter Folder Name"
            type="text"
          />
        </div>
        <div
          className="flex justify--space-around"
        >
          <button
            className="File__btn--round File__btn--cancel"
            onClick={[Function]}
          />
          <button
            className="File__btn--round File__btn--add"
            onClick={[Function]}
          />
        </div>
      </div>
    </div>
    <div
      className="FileBrowser__empty"
    >
      <h5>
        Upload Files by Dragging & Dropping Here
      </h5>
    </div>
  </div>
</div>
`;
